0002-RCBS-Crypto-building-block.md

```diff
@@ -0,0 +1,675 @@
+
+Algorithms available will depend on what the backend vaults support, but in general developers should always find AES (encrypt/decrypt only) and RSA; when supported, we can offer also ChaCha20-Poly1305 (encrypt/decrypt only) and ECC with ECDSA or EdDSA (sign/verify only).
+
+## Related Items
+
+Previous proposal as GitHub issue: dapr/dapr#4508
+
+## Data flow: runtime and components
+
+![data flow](./resources/0002-RCBS-Crypto-building-block/data-flow.png)
+
+## Dapr encryption scheme: dapr.io/enc/v1
+
+In the first version of the building block, we define 2 higher level operations to encrypt and decrypt data, in addition to low-level operations.
+
+> **Sources:** The encryption scheme that Dapr uses is heavily inspired by the [Tink wire format](https://developers.google.com/tink/wire-format) (from the Tink library maintained by Google), as well as by Filippo Valsorda's [age](https://age-encryption.org/v1), and Minio's [DARE](https://github.com/minio/sio).
+
+The Dapr encryption scheme is optimized for processing data as a stream. Data is chunked into multiple parts which are encrypted independently. This allows us to return data to callers as a stream, even when decrypting messages, being confident that we are not flushing unverified data to the client.
+
+### Key
+
+Each message is encrypted with a 256-bit symmetric **File Key (FK)** that is randomly generated by Dapr for each new message. The key must be generated as 32 byte of output from a CSPRNG (such as Go's `crypto/rand.Reader`) and must not be reused for other files.
+
+The FK is wrapped using a key stored in a key vault (**Key Encryption Key (KEK)**) by Dapr. The result of the wrapping operation is the **Wrapped File Key (WFK)**.  The algorithm used depends on the type of the KEK as well as the algorithms supported by the component: in order of preference:
+
+- Symmetric keys:
+    - AES-KW ([RFC 3394](https://www.rfc-editor.org/rfc/rfc3394.html)): `AES-KW`
+    - AES-GCM-SIV: `AES-GCM-SIV`  
+      This will be implemented if it can be used with keys stored in key vaults; to be confirmed.
+- RSA keys:
+    - RSA OAEP with SHA-256: `RSA-OAEP-256`
+
+> In the future, we should explore how to add support for elliptic curve cryptography, for example P-256/P-384/P-521 or Curve25519, which requires performing a static ECDH key agreement.
+
+### Ciphertext format
+
+The ciphertext is formatted as:
+
+```text
+header || binary payload
+```
+
+### Header
+
+The **header** is human-readable and contains 3 items, each terminated by a line feed (0x0A) character:
+
+1. Name and version of the encryption scheme used. Currently, this is always `dapr.io/enc/v1`
+2. The manifest, which is a JSON object.
+3. The MAC for the header, base64-encoded
```

Comment:
So, it should be easy to add more to the metadata by appending and using the line feed as terminator, right?

Reply:
Additional metadata could be added to the manifest, which is a JSON object. However, changes to the metadata would require a v2 of the protocol.

Any metadata in particular you are concerned with?